src/components/users/controllers.tsx

```diff
@@ -95,3 +106,208 @@ export async function getUser(
+      status: 404,
+    };
+  }
+
+  const uaaUser = await uaa.getUser(accountsUser.uuid)
+  if (!uaaUser) {
+    // 100ms delay to stop people iterating through users
+    await new Promise(finish => setTimeout(finish, 100));
+
+    template.title = 'Error: User not found';
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} userNotFound={true} />),
+      status: 404,
+    };
+  }
+
+  if (uaaUser.origin != 'uaa') {
+    // 100ms delay to stop people iterating through users
+    await new Promise(finish => setTimeout(finish, 100));
+
+    const idpNice = capitalize(uaaUser.origin);
+    template.title = `Error: You have enabled ${idpNice} single sign-on`;
+
+    return {
+      body: template.render(<PasswordResetRequest csrf={ctx.viewContext.csrf} userEnabledSSO={true} idpNice={idpNice}/>),
+    };
+  }
+
+  const notify = new NotificationClient({
+    apiKey: ctx.app.notifyAPIKey,
+    templates: {
+      passwordReset: ctx.app.notifyPasswordResetTemplateID,
+    },
+  });
+  const resetCode = await uaa.obtainPasswordResetCode(uaaUser.userName);
+
+  const url = new URL(ctx.app.domainName);
+  url.pathname = '/password/confirm-reset';
+  url.searchParams.set('code', resetCode);
+
+  await notify.sendPasswordReminder(email, url.toString());
+
+  template.title = 'Password reset successfully requested';
+  const successPage = <PasswordResetSuccess
+    title="Password reset successfully requested"
+    message="You should receive an email shortly with further instructions."
+  />;
+
+  return {
+    body: template.render(successPage),
+  };
+}
+
+export async function resetPasswordProvideNew(ctx: IContext, params: IParameters): Promise<IResponse> {
+  if (!params.code) {
+    throw new NotFoundError('Invalid password reset token.');
+  }
+
+  const template = new Template(ctx.viewContext, 'Password reset');
+
+  return await Promise.resolve({
+    body: template.render(<PasswordResetSetPasswordForm csrf={ctx.viewContext.csrf} code={params.code} />),
+  });
+}
+
+export function checkPasswordAgainstPolicy(pw: string): { valid: boolean; message?: string } {
+  if (pw.length < 12) {
+    return {
+      valid: false,
+      message: 'Your password should be 12 characters or more',
+    }
+  }
+
+  if (!pw.match(/[a-z]/)) {
+    return {
+      valid: false,
+      message: 'Your password should contain a lowercase character',
+    }
+  }
+
+  if (!pw.match(/[A-Z]/)) {
+    return {
+      valid: false,
+      message: 'Your password should contain an uppercase character',
+    }
+  }
+
+  if (!pw.match(/[0-9]/)) {
+    return {
+      valid: false,
+      message: 'Your password should contain a number',
+    }
+  }
+
+  if (!pw.match(/[-_+:;<>[\]()#@£%^&!/]/)) {
+    return {
+      valid: false,
+      message: 'Your password should contain a special character',
+    }
+  }
+
+  return { valid: true };
```

Comment:
I'm not a biggest fan of this function. It's independent from UAA, meaning we may be passing on one front and failing on another (at some point in time).

It could also lead to frustrated users, just adding random characters without memorising the password causing another need for password reset.

Just saying, no need to act :)

Reply:
The password policy is more restrictive than what we have in UAA

I think we should make UAA more restrictive to match this password policy

Reply:
I'm not saying it isn't. I mean it in a way, that if there is a weird bug or misconfiguration of UAA to require, say `At least three digits in a password`, our filter could pass it through whilst UAA would reject it.

I understand there is little to no chance of that happening, but if it would, it'd be a pain to investigate for someone who wasn't involved in this review 🤔 

Once again, ignore me - I'm happy with it if you are :)

Reply:
You raise a good point, we aren't returning the UAA error to the user if UAA rejects the `/password_change` POST, although we will return an error to the user because of how the UAA client works.

I think it is a sufficiently rare case that I'm fine with it